### PR TITLE
[CTF Rebuild] Added repairing of blob-blocks (doors, platforms, bridges)

### PR DESCRIPTION
**Pull request by bunniewormy**
## Status

(pick one)

- **READY**: this PR is (to the best of your knowledge) ready to be incorporated into the game.

## Description

[fix] current sprite frame of blob-blocks like doors, platforms, etc. is now based on health rather than amount of hits received
[added] you can repair blob-blocks same as you repair normal blocks. This is done by "healing" them. Takes the same amount of materials as if you wanted to build a new one.

### CODE REVIEW

**Comment by MonkeyFeats** [comment]
This will encourage even more dull 'rat' gameplay.

**Comment by bunniewormy** [comment]
> 
> 
> This will encourage even more dull 'rat' gameplay.

This PR, same as other rebuild PRs, should be thought of as a part of a bigger rebalance. The offense/defense changes rely on each other a lot.

The main argument for this is consistency - it doesn't really make sense that you can't repair blob-blocks like normal stone or wood blocks. How will this result in "even more dull rat gameplay", especially when you combine this with other rebuild changes?

File: Entities/Common/Building/BlobPlacement.as [hunk 1]
```diff
@@ -42,7 +50,7 @@ void PlaceBlob(CBlob@ this, CBlob @blob, Vec2f cursorPos)
 }
 
 // Returns true if pos is valid
-bool serverBlobCheck(CBlob@ blob, CBlob@ blobToPlace, Vec2f cursorPos)
+bool serverBlobCheck(CBlob@ blob, CBlob@ blobToPlace, Vec2f cursorPos, bool repairing=false)
```
Comment (Vam-Jam) [inline Entities/Common/Building/BlobPlacement.as:53]: space = space false

File: Entities/Common/Building/BlobPlacement.as [hunk 2]
```diff
@@ -346,10 +355,23 @@ void onTick(CBlob@ this)
...
+						params.write_Vec2f(getBottomOfCursor(bc.tileAimPos, carryBlob));
+						this.SendCommand(this.getCommandID("repairBlob"), params);
+					}
+					else
+					{
+						CBitStream params;
+						params.write_u16(carryBlob.getNetworkID());
+						params.write_Vec2f(getBottomOfCursor(bc.tileAimPos, carryBlob));
+						this.SendCommand(this.getCommandID("placeBlob"), params);
+					}
```
Comment (Vam-Jam) [inline Entities/Common/Building/BlobPlacement.as:373]: could write to params before the if statement, and just send a different cmd based on the statement, less code dupe

File: Entities/Common/Building/PlacementCommon.as [hunk 3]
```diff
@@ -126,25 +126,40 @@ bool isBuildableAtPos(CBlob@ this, Vec2f p, TileType buildTile, CBlob @blob, boo
...
-		if (!isSeed && !isLadder && (buildSolid || isSpikes || isDoor) && map.getSectorAtPosition(middle, "no build") !is null)
+		s32 x = Maths::Floor(p.x);
+		x /= map.tilesize;
+		s32 y = Maths::Floor(p.y);
+		y /= map.tilesize;
+
+		CBlob@ blobAtPos = map.getBlobAtPosition(Vec2f(p.x, p.y));
+
+		// repairing blobs
+		if(blobAtPos !is null && blob !is null && (isDoor || isPlatform))
```
Comment (Vam-Jam) [inline Entities/Common/Building/PlacementCommon.as:152]: if space (

File: Entities/Common/Building/PlacementCommon.as [hunk 4]
```diff
@@ -126,25 +126,40 @@ bool isBuildableAtPos(CBlob@ this, Vec2f p, TileType buildTile, CBlob @blob, boo
...
 		{
-			return false;
+			if(blobAtPos.getName() == blob.getName() &&
```
Comment (Vam-Jam) [inline Entities/Common/Building/PlacementCommon.as:154]: if space (

File: Entities/Common/Building/PlacementCommon.as [hunk 5]
```diff
@@ -126,25 +126,40 @@ bool isBuildableAtPos(CBlob@ this, Vec2f p, TileType buildTile, CBlob @blob, boo
...
 
 		Vec2f middle = p;
 
-		if (!isSeed && !isLadder && (buildSolid || isSpikes || isDoor) && map.getSectorAtPosition(middle, "no build") !is null)
+		s32 x = Maths::Floor(p.x);
+		x /= map.tilesize;
+		s32 y = Maths::Floor(p.y);
+		y /= map.tilesize;
+
+		CBlob@ blobAtPos = map.getBlobAtPosition(Vec2f(p.x, p.y));
```
Comment (Vam-Jam) [inline Entities/Common/Building/PlacementCommon.as:149]: would it not be better to get all blobs at said position in case we have overlapping blobs, it could prevent you from repairing

File: Entities/Structures/Platform/Bridge.as [hunk 6]
```diff
@@ -36,16 +36,23 @@ void onInit(CBlob@ this)
...
 void onHealthChange(CBlob@ this, f32 oldHealth)
 {
-	MakeDamageFrame(this);
+	f32 hp = this.getHealth();
+	bool repaired = hp > oldHealth ? true : false;
+	MakeDamageFrame(this, repaired);
 }
 
-void MakeDamageFrame(CBlob@ this)
+void MakeDamageFrame(CBlob@ this, bool repaired=false)
```
Comment (Vam-Jam) [inline Entities/Structures/Platform/Bridge.as:44]: repaired space = space false

File: Entities/Structures/Platform/Bridge.as [hunk 7]
```diff
@@ -36,16 +36,23 @@ void onInit(CBlob@ this)
...
 {
 	f32 hp = this.getHealth();
 	f32 full_hp = this.getInitialHealth();
 	int frame_count = this.getSprite().animation.getFramesCount();
 	int frame = frame_count - hp / full_hp * frame_count;
 	this.getSprite().animation.frame = frame;
+
+	if(repaired)
```
Comment (Vam-Jam) [inline Entities/Structures/Platform/Bridge.as:52]: if space (

File: Entities/Structures/Platform/WoodenPlatform.as [hunk 8]
```diff
@@ -28,6 +28,29 @@ void onInit(CBlob@ this)
...
+
+void MakeDamageFrame(CBlob@ this, bool repaired=false)
+{
+	f32 hp = this.getHealth();
+	f32 full_hp = this.getInitialHealth();
+	int frame_count = this.getSprite().animation.getFramesCount();
+	int frame = frame_count - hp / full_hp * frame_count;
+	this.getSprite().animation.frame = frame;
+
+	if(repaired)
```
Comment (Vam-Jam) [inline Entities/Structures/Platform/WoodenPlatform.as:50]: if space (

File: Entities/Common/Building/BlobPlacement.as (hunk 1 above)
Comment (bunniewormy) [reply Entities/Common/Building/BlobPlacement.as]: fixed

File: Entities/Common/Building/BlobPlacement.as (hunk 2 above)
Comment (bunniewormy) [reply Entities/Common/Building/BlobPlacement.as:372]: fixed

File: Entities/Common/Building/PlacementCommon.as (hunk 4 above)
Comment (bunniewormy) [reply Entities/Common/Building/PlacementCommon.as]: fixed

File: Entities/Common/Building/PlacementCommon.as (hunk 3 above)
Comment (bunniewormy) [reply Entities/Common/Building/PlacementCommon.as]: fixed

File: Entities/Common/Building/PlacementCommon.as (hunk 5 above)
Comment (bunniewormy) [reply Entities/Common/Building/PlacementCommon.as]: very true, fixed

File: Entities/Structures/Platform/Bridge.as (hunk 6 above)
Comment (bunniewormy) [reply Entities/Structures/Platform/Bridge.as]: fixed

File: Entities/Structures/Platform/Bridge.as (hunk 7 above)
Comment (bunniewormy) [reply Entities/Structures/Platform/Bridge.as]: fixed

File: Entities/Structures/Platform/WoodenPlatform.as (hunk 8 above)
Comment (bunniewormy) [reply Entities/Structures/Platform/WoodenPlatform.as]: fixed

File: Entities/Structures/Door/SwingDoor.as [hunk 9]
```diff
@@ -187,53 +190,52 @@ f32 onHit(CBlob@ this, Vec2f worldPoint, Vec2f velocity, f32 damage, CBlob@ hitt
...
 
-	CSprite @sprite = this.getSprite();
+	return damage;
+}
 
-	if (sprite !is null)
+void onHealthChange(CBlob@ this, f32 oldHealth)
+{
+	f32 hp = this.getHealth();
+	bool repaired = hp > oldHealth ? true : false;
```
Comment (AsuMagic) [inline Entities/Structures/Door/SwingDoor.as:199]: `bool repaired = (hp > oldHealth);` is semantically identical

File: Entities/Structures/Door/SwingDoor.as [hunk 10]
```diff
@@ -149,6 +149,10 @@ void onCollision(CBlob@ this, CBlob@ blob, bool solid)
 	if (blob !is null)
 	{
 		this.getCurrentScript().tickFrequency = 3;
+		if (!isOpen(this) && canOpenDoor(this, blob))
```
Comment (AsuMagic) [inline Entities/Structures/Door/SwingDoor.as:152]: Why is this needed here and what consequences does it have vs the current behavior?  
Is it just a cleanup of what is otherwise done in `doesCollideWithBlob`?

File: Entities/Common/Building/BlobPlacement.as [hunk 11]
```diff
@@ -464,6 +485,19 @@ void onCommand(CBlob@ this, u8 cmd, CBitStream @params)
 			SendGameplayEvent(createBuiltBlobEvent(this.getPlayer(), carryBlob.getName()));
 		}
 	}
+	if (cmd == this.getCommandID("repairBlob"))
+	{
+		CBlob @carryBlob = getBlobByNetworkID(params.read_u16());
+		Vec2f pos = params.read_Vec2f();
+
+		CBlob @repairBlob = getMap().getBlobAtPosition(pos);
```
Comment (AsuMagic) [inline Entities/Common/Building/BlobPlacement.as:493]: I think it would be better to send the network ID of the repaired blob tile as part of the bitstream rather than figuring out from the serverside. I could imagine it doing weird stuff in some odd cases especially with lag.

e.g. what happens if a 300ms ping builder sends a `repairBlob` command and by the time it's processed serverside, the tile being repaired was broken and _something else entirely_ than a tile now overlaps `pos`? Even if it might be nothing it's making the whole thing more obscure

Comment (AsuMagic) [reply Entities/Common/Building/BlobPlacement.as:493]: Related to this command, if a laggy builder holds LMB to repair a door, but that door gets broken by the time the command reaches the server, should we make it place a new blob instead?

**Review by AsuMagic — CHANGES_REQUESTED** [review]

File: Entities/Structures/Platform/Bridge.as [hunk 12]
```diff
@@ -36,16 +36,23 @@ void onInit(CBlob@ this)
 
 void onHealthChange(CBlob@ this, f32 oldHealth)
 {
-	MakeDamageFrame(this);
+	f32 hp = this.getHealth();
+	bool repaired = hp > oldHealth ? true : false;
```
[the remark this answers is not in the document]
Comment (bunniewormy) [reply Entities/Structures/Platform/Bridge.as]: derp, fixed

File: Entities/Structures/Door/SwingDoor.as (hunk 9 above)
Comment (bunniewormy) [reply Entities/Structures/Door/SwingDoor.as]: derp, fixed

File: Entities/Structures/Platform/WoodenPlatform.as [hunk 13]
```diff
@@ -28,6 +28,29 @@ void onInit(CBlob@ this)
...
+	bool repaired = hp > oldHealth ? true : false;
+	MakeDamageFrame(this, repaired);
+}
+
+void MakeDamageFrame(CBlob@ this, bool repaired = false)
+{
+	f32 hp = this.getHealth();
+	f32 full_hp = this.getInitialHealth();
+	int frame_count = this.getSprite().animation.getFramesCount();
+	int frame = frame_count - hp / full_hp * frame_count;
```
[the remark this answers is not in the document]
Comment (bunniewormy) [reply Entities/Structures/Platform/WoodenPlatform.as]: fixed for all 3 files

File: Entities/Structures/Door/SwingDoor.as (hunk 10 above)
Comment (bunniewormy) [reply Entities/Structures/Door/SwingDoor.as:152]: yes, I removed that part from doesCollideWithBlob and put it here - more logical this way imo

File: Entities/Common/Building/BlobPlacement.as (hunk 11 above)
Comment (bunniewormy) [reply Entities/Common/Building/BlobPlacement.as:520]: > e.g. what happens if a 300ms ping builder sends a repairBlob command and by the time it's processed serverside, the tile being repaired was broken and something else entirely than a tile now overlaps pos? Even if it might be nothing it's making the whole thing more obscure

would probably end up passing, i've added additional checks in serverBlobCheck now

> Related to this command, if a laggy builder holds LMB to repair a door, but that door gets broken by the time the command reaches the server, should we make it place a new blob instead?

probably should - added that functionality, but either way it requires sending the pos instead of network id 

also, i've added some getHealth() checks along with null checks in certain places because apparently a blob can exist without being null (exist with negative health for a moment and not show up), which meant that if you tried placing a blobtile right after destroying it, it wouldn't place

**Review by AsuMagic — DISMISSED** [review]

**Review by AsuMagic — APPROVED** [review]
